keps/sig-storage/20200120-generic-data-populators.md

```diff
@@ -0,0 +1,309 @@
+to be pre-populated with various OS images. OS images tend to be stored in external
+repositories dedicated to that purpose, often with various mechanisms for retrieving
+them efficiently that are external to Kubernetes.
+
+One way to achieve this is to represent disk images as custom resources that point to
+the image repository, and to allow creation of PVCs from these custom resources such
+that the volumes come pre-populated with the correct data. Efficient population of the
+data could be left up to a purpose-built controller that knows how to get the bits
+where they need to be with minimal I/O.
+
+#### Backup/Restore
+
+Without getting into the details of how backup/restore should be implemented, it's
+clear that whatever design one chooses, a necessary step is to have the user
+(or higher level controller) create a PVC that points to the backup they want to
+restore, and have the data appear in that volume somehow.
+
+One can imagine backups simply being a special case of snapshots, in which case the
+existing design is sufficient, but if you want anything more sophisticated, there
+will inevitably be a new object type that represents a backup. While it's arguable
+that backup should be something CSI plugins should be exclusively responsible for,
+one can also argue that generic backup tools should also exist which can backup
+and restore all kind of volumes. Those tools will be apart from CSI plugins and
+yet need a way to populate volumes with restored volumes.
+
+It's also likely that multiple backup/restore implementations will be developed,
+and it's not a good idea to pick a winner at the Kubernetes API layer. It makes
+more sense to enable developers to try different approaches by making the API allow
+restoring from various kinds of things. 
+
+### Implementation Details/Notes/Constraints
+
+As noted above, the proposal is extremely simple -- just remove the validation on
+the `DataSource` field. This raises the question of WHAT will happen when users
+put new things in that field, and HOW populators will actually work with so small
+a change.
+
+It's first important to note that only consumers of the `DataSource` field today
+are the various dynamic provisioners, most notably the external-provisioner CSI
+sidecar. If the external-provisioner sidecar sees a data source it doesn't
+understand, it simply ignores the request, which is both important for forward
+compatibility, and also perfect for the purposes of a data populator. This allows
+developers to add new types of data sources that the dynamic provisioners will
+simply ignore, enabling a different controller to see these objects and respond
+to them.
+
+The way I expect most data populators to work in practice is that they will see
+a request for a PVC `foo` with a data source of `bar`. The populator will create
+a shadow PVC `foo'` in another namespace with all of the same spec except the
```

Comment:
If another namespace is involved, should this be using the method provided by the namespace transfer KEP?

Reply:
Namespace transfer is already easy to do if you have a controller and the appropriate cluster role/RBAC. The example controller demonstrates this.

The purpose of the namespace-transfer feature (as I understand it) is to provide the same functionality to low-privilege users through some kind of standardized interface.

Conceivably, if the namespace transfer feature is added, then we could use that capability to simplify the design of populators, and require less privileges. However it's not required.

Reply:
Yes the idea is that we don't want end users to see weird PVCs and pods appearing in their namespace, but we need to create at least a pod to do the population work. My proposal also involves a second PVC to work around the problem that we don't want and end-user's pod to race with the populator pod to attach to the PVC first. By doing the population in a different PVC, we avoid that problem. The second namespace serves to just hide what's going on from the end user (assuming the end user can't see all namespaces). The re-binding of the PV is that hardest part of the scheme, but once you have a way to rebind PVs, then moving them across namespaces is trivial because the PV itself is not-namespaced. I will do a separate writeup of PV-rebinding techniques since it seems interesting for a variety of reasons.